bakery/checker.go

```diff
@@ -0,0 +1,551 @@
+// hold whether its respective operation was allowed.
+//
+// If all the operations succeeded, the returned slice will be nil.
+//
+// The returned *AuthInfo will always be non-nil.
+//
+// The LoginOp operation is treated specially - it is always required if
+// present in ops.
+func (a *AuthChecker) AllowAny(ctxt context.Context, ops ...Op) (*AuthInfo, []bool, error) {
+	authed, used, err := a.allowAny(ctxt, ops)
+	return a.newAuthInfo(used), authed, err
+}
+
+func (a *AuthChecker) newAuthInfo(used []bool) *AuthInfo {
+	info := &AuthInfo{
+		Identity:  a.identity,
+		Macaroons: make([]macaroon.Slice, 0, len(a.macaroons)),
+	}
+	for i, isUsed := range used {
+		if isUsed {
+			info.Macaroons = append(info.Macaroons, a.macaroons[i])
+		}
+	}
+	return info
+}
+
+// allowAny is the internal version of AllowAny. Instead of returning an
+// authInfo struct, it returns a slice describing which operations have
+// been successfully authorized and a slice describing which macaroons
+// have been used in the authorization.
+func (a *AuthChecker) allowAny(ctxt context.Context, ops []Op) (authed, used []bool, err error) {
+	if err := a.init(ctxt); err != nil {
+		return nil, nil, errgo.Mask(err)
+	}
+	logger.Infof("after authorizer init, identity %#v", a.identity)
+	used = make([]bool, len(a.macaroons))
+	authed = make([]bool, len(ops))
+	numAuthed := 0
+	var errors []error
+	for i, op := range ops {
+		for _, mindex := range a.authIndexes[op] {
+			_, err := a.checkConditions(ctxt, op, a.conditions[mindex])
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			authed[i] = true
+			numAuthed++
+			used[mindex] = true
+			break
+		}
+	}
+	if a.identity != nil {
+		// We've authenticated as a user, so even if the operations didn't
+		// specifically require it, we add the login macaroon
+		// to the macaroons used.
+		// Note that the LoginOp conditions have already been checked
+		// successfully in initOnceFunc so no need to check again.
+		// Note also that there may not be any macaroons if the
+		// identity client decided on an identity even with no
+		// macaroons.
+		for _, i := range a.authIndexes[LoginOp] {
+			used[i] = true
+		}
+	}
+	if numAuthed == len(ops) {
+		// All operations allowed.
+		return nil, used, nil
+	}
+	// There are some unauthorized operations.
+	need := make([]Op, 0, len(ops)-numAuthed)
+	needIndex := make([]int, cap(need))
+	for i, ok := range authed {
+		if !ok {
+			needIndex[len(need)] = i
+			need = append(need, ops[i])
+		}
+	}
+	logger.Infof("operations needed after authz macaroons: %#v", need)
+
+	// Try to authorize the operations even even if we haven't got an authenticated user.
+	oks, caveats, err := a.p.Authorizer.Authorize(ctxt, a.identity, need)
+	if err != nil {
+		return authed, used, errgo.Notef(err, "cannot check permissions")
+	}
+	if len(oks) != len(need) {
+		return authed, used, errgo.Newf("unexpected slice length returned from Allow (got %d; want %d)", len(oks), len(need))
```

Comment:
I dunno. I don't want callers checking for it particularly. It's almost a panic.

I've changed the definition of Authorize so that this check isn't necessary
(additional elements are treated as false).